service/collector/token_expiration.go

```diff
@@ -0,0 +1,109 @@
+package collector
+
+import (
+	"context"
+	"strconv"
+
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+	"github.com/prometheus/client_golang/prometheus"
+	"k8s.io/client-go/kubernetes"
+)
+
+const (
+	labelSubscriptionId  = "subscription_id"
+	labelApplicationId   = "application_id"
+	labelApplicationName = "application_name"
+	labelExpirationTS    = "expiration_ts"
+)
+
+var (
+	tokenExpirationDesc *prometheus.Desc = prometheus.NewDesc(
+		prometheus.BuildFQName("azure_operator", "service_principal_token", "expiration"),
+		"Expiration date for Azure Access Tokens.",
+		[]string{
+			labelSubscriptionId,
+			labelApplicationId,
+			labelApplicationName,
+			labelExpirationTS,
+		},
+		nil,
+	)
+)
+
+type ServicePrincipalTokenConfig struct {
+	K8sClient kubernetes.Interface
+	Logger    micrologger.Logger
+}
+
+type ServicePrincipalToken struct {
+	k8sClient kubernetes.Interface
+	logger    micrologger.Logger
+
+	environmentName string
+}
+
+func NewServicePrincipalToken(config ServicePrincipalTokenConfig) (*ServicePrincipalToken, error) {
+	if config.K8sClient == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.K8sClient must not be empty", config)
+	}
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Logger must not be empty", config)
+	}
+
+	v := &ServicePrincipalToken{
+		k8sClient: config.K8sClient,
+		logger:    config.Logger,
+	}
+
+	return v, nil
+}
+
+func (spt *ServicePrincipalToken) Collect(ch chan<- prometheus.Metric) error {
+	ctx := context.Background()
+
+	clientSets, err := getClientSets(spt.k8sClient, spt.environmentName)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	for _, item := range clientSets {
+		clientSet := item
+
+		applications, err := clientSet.ApplicationsClient.ListComplete(ctx, "")
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		for applications.NotDone() {
+			app := applications.Value()
+
+			if *app.AppID == clientSet.ClientID {
+				for _, pc := range *app.PasswordCredentials {
+					ch <- prometheus.MustNewConstMetric(
+						tokenExpirationDesc,
```

Comment:
I am curious: is there a better way of modeling this metric? I know that a _date_ it's a weird metric. But it also feels weird to have a gauge to always be 1, labeled with a date. Will we be able to alert on this?

Reply:
Maybe the gauge can be the days until the certificate expires so that the alert will be triggered when the days are lower than certain threshold?

Reply:
https://www.robustperception.io/get-alerted-before-your-ssl-certificates-expire

Reply:
We use the approach of storing the unix TS on similar things, such as the certificate expiration: https://prometheus.g8s.talos.westeurope.azure.gigantic.io/graph?g0.range_input=1h&g0.expr=cert_exporter_not_after&g0.tab=1

I'd stick with the same approach here as well, since we're talking about similar metrics

Reply:
But I think in that other approach it's storing expiration date in unix timestamp as a value, not label, but here it seems to be label. My intuition would be to store value of `(pc.EndDate.Unix() - time.Now().Unix())` so you could simply alert when the value is close to zero. It makes alert function a lot more readable while the measurement is also pretty readable. Nevertheless, either way. Your call :slightly_smiling_face: